src/app/clusters/ota-requestor/OTARequestor.cpp

```diff
@@ -450,6 +448,69 @@ void OTARequestor::NotifyUpdateApplied(uint32_t version)
     ConnectToProvider(kNotifyUpdateApplied);
 }
 
+CHIP_ERROR OTARequestor::GetDefaultOtaProviderList(AttributeValueEncoder & encoder)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    return encoder.EncodeList([&](const auto & enc) -> CHIP_ERROR {
+        mDefaultOtaProviderList.ForEachActiveObject([&](ProviderLocation::Type * pl) {
+            err = enc.Encode(*pl);
+            VerifyOrReturnError(err == CHIP_NO_ERROR, Loop::Break);
+            return Loop::Continue;
+        });
+        return err;
+    });
+}
+
+CHIP_ERROR OTARequestor::ClearDefaultOtaProviderList(FabricIndex fabricIndex)
+{
+    // Remove all entries for the fabric index indicated
+    mDefaultOtaProviderList.ForEachActiveObject([&](ProviderLocation::Type * pl) {
+        if (pl->GetFabricIndex() == fabricIndex)
+        {
+            mDefaultOtaProviderList.ReleaseObject(pl);
+        }
+        return Loop::Continue;
+    });
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR OTARequestor::AddDefaultOtaProvider(FabricIndex fabricIndex, ProviderLocation::Type const & providerLocation)
+{
+    if (providerLocation.GetFabricIndex() == kUndefinedFabricIndex)
+    {
+        return CHIP_IM_GLOBAL_STATUS(ConstraintError);
+    }
+
+    // Look for an entry with the same fabric index indicated
+    bool fabricIndexExists = false;
+    mDefaultOtaProviderList.ForEachActiveObject([&](auto * pl) {
+        if (pl->GetFabricIndex() == fabricIndex)
+        {
+            fabricIndexExists = true;
+            return Loop::Break;
+        }
+        return Loop::Continue;
+    });
+
+    // Check if there is already an entry with the same fabric index
+    if (fabricIndexExists)
```

Comment:
Wouldn't it be better to just overwrite the entry? We are already saying that we only support ReplaceAll.

Reply:
That would not be per-spec, right?

In particular, if a client tries to write a list with two entries in it, the spec behavior is to have an error, not to just silently pretend like the first entry was not there.

Reply:
Sure, writing a list with two entries for the same fabric is invalid but is this the place to detect it? My concern is attempting to write a valid list while a fabric entry already exists in the cluster. Should the whole write operation fail just because one entry already exists?
Or will the write succeed for the rest of the entries on the list but not this one? (Which is also kinda counter-intuitive)

Reply:
Current behavior will succeed until an error is encountered (so a partial write could occur leading up to the error). I didn't want to put too much logic here but if we really wanted to, perhaps the right behavior should be to wipe out the valid entries that were written already. This would enforce that a successful write could only occur if the client passed in only valid values (in this case, only one entry). I'm open to this if you guys feel this is necessary.

Reply:
Is there a way to detect duplicates in the input list earlier in the processing? That way there's no need to wipe out anything. Also that way we can silently overwrite the existing entries instead of failing (which is the case I was concerned about here).

Reply:
But I don't think any of this is hugely important, I'm OK with any choice

Reply:
> Is there a way to detect duplicates in the input list earlier in the processing?

Not easily.  This code is given the list items one item at a time...

Reply:
Keeping the code as-is. Currently, we can't write more than one entry for the same accessing fabric index anyway. Any additional entries would be ignored (and an error message logged to indicate so).